### PR TITLE
Add interface package with sidebar functionality

### PR DESCRIPTION
**Pull request by jorgefilipecosta**
## Description
Supersedes: https://github.com/WordPress/gutenberg/pull/20336
This PR starts an admin-screen package suggested by @youknowriad in https://github.com/WordPress/gutenberg/pull/20336#pullrequestreview-369397086.

This package for now just contains the sidebar functionality but other functionality will be added (e.g: fullscreen mode).

The sidebar functionality is renamed the ComplementaryArea as suggested by @gziolo.
This package, for now, is experimental should be private and is bundled with other packages.

Regarding the admin screen store:

We implement an API in the store that allows keeping track of which area is active in a given zone when only one area can be active at a time. For example what sidebar is being rendered, what tab in a tab system is enabled etc...

```
wp.data.dispatch('core/admin-screen').setSingleActiveArea( 'edit-site/complementary-area', 'edit-site/block-inspector'  );

wp.data.select('core/admin-screen').getSingleActiveArea( 'edit-site/complementary-area' ); -> "edit-site/block-inspector"

wp.data.dispatch('core/admin-screen').setSingleActiveArea( 'edit-site/complementary-area', 'edit-site/global-styles'  );

wp.data.select('core/plugins').getSingleActiveArea( 'edit-site/complementary-area' ); -> "edit-site/global-styles"

wp.data.dispatch('core/plugins').setSingleActiveArea( 'edit-site/complementary-area' );

wp.data.select('core/plugins').getSingleActiveArea( 'edit-site/complementary-area' ); -> undefined
```

We implement an API in the store that allows keeping track of which areas are active in a given zone when multiple areas can be active at a time
E.g: which plugins are pinned, which panels are open.
This API allows plugins to keep track for example which panels are open and take advantage of the persisting mechanism (when we add it) without the need to implement their own store/persistence.

```
wp.data.select('core/admin-screen').isMultipleActiveAreaActive( 'edit-post/complementary-area', 'edit-post/block-patterns-sidebar' ); -> false

wp.data.dispatch('core/admin-screen').setMultipleActiveAreaEnableState( 'edit-post/complementary-area', 'edit-post/block-patterns-sidebar', true );

wp.data.select('core/admin-screen').isMultipleActiveAreaActive( 'edit-post/complementary-area', 'edit-post/block-patterns-sidebar' ); -> true
```

## How has this been tested?
I verified the edit-post sidebar still works as expected (namely auto-closing on mobile, keyboard shortcuts, plugin sidebar).
I verified the sidebar mechanism added on edit-site works as expected.

### CODE REVIEW

**Comment by mtias** [comment]
`admin-screen` seems like a weird name for this to me.

**Comment by youknowriad** [comment]
@mtias we have different things we want to reuse across screens. Sidebar Plugins, EditorSkeleton (PageSkeleton), FullscreenMode 

What do you suggest as a name?

**Comment by mtias** [comment]
I know it's already in use, but this would be the "editor" package to me (as opposed to the block editor).

**Comment by jorgefilipecosta** [comment]
Maybe editor-screen would be a good name?

**Comment by mtias** [comment]
Can you list all the "editor" related packages we'd have?

**Comment by jorgefilipecosta** [comment]
> Can you list all the "editor" related packages we'd have?

Hi @mtias, 
We would have "block-editor", "editor" and "editor-screen" plus three packages that would potentially use editor-screen: "edit-post", "edit-site", "edit-widgets".

**Comment by mtias** [comment]
What are the conceptual differences among `block-editor`, `editor` and `editor-screen`?

**Comment by jorgefilipecosta** [comment]
> What are the conceptual differences among block-editor, editor and editor-screen?

Hi @mtias,
- <b>block-editor</b> -> Generic block editor, responsible for dealing with the blocks. Given a data structure that represents a set of blocks renders a UI to edit these blocks, to remove, move, and insert new blocks. Not WordPress specific.
- <b>editor</b> -> Provides a set of helpers to deal with WordPress posts and its edits. I think today we could do almost everything this module does with our data APIs; in the long term, this module should probably be deprecated.
- <b>editor-screen</b> -> Provides a set of helpers to build a screen UI in WordPress similar to the post edit screen Gutenberg offers. Allows controlling which areas are active or not, e.g., block-inspector, a plugin sidebar, what modes are active, e.g., fullscreen., what items are open or closed the panels, and deals with persisting these settings.

**Comment by mtias** [comment]
`editor` and `editor-screen` should likely be folded together then.

**Comment by youknowriad** [comment]
Do you think these components/tools (Skeleton, Sidebar, Fullscreen) are things specific to editor screens or any Admin screen? Personally, I'm thinking the latter.

**Comment by gziolo** [comment]
Whatever the final decision is, we should improve the documentation for all mentioned packages and provide a high-level overview of how they all fit together. It's always been an issue for all people that start with Gutenberg development.

**Comment by jorgefilipecosta** [comment]
> Do you think these components/tools (Skeleton, Sidebar, Fullscreen) are things specific to editor screens or any Admin screen? Personally, I'm thinking the latter.

Good point, I agree these UI artifacts may be the basis to a screen in WordPress that may not necessarily contain an editor, e.g: I may imagine this artifact may contain a system to manage tickets, and the sidebars show information about the person that submitted the ticket.

What do you think @mtias?

**Comment by chrisvanpatten** [comment]
I agree with @youknowriad and @jorgefilipecosta — we're already brainstorming moving some non-editor UI over to React screens driven by these components.

Perhaps a very generic name like `@wordpress/interface` would be an option? Or `@wordpress/layout`?

**Comment by mtias** [comment]
`@wordpress/interface` seems fine as it clarifies a bit the relationship with `@wordpress/components` (components _make_ the interface).

**Comment by jorgefilipecosta** [comment]
This PR was updated it now proposes an interface package as suggested by @mtias. cc: @youknowriad

File: packages/edit-post/src/store/reducer.js [hunk 1]
```diff
@@ -111,19 +102,6 @@ export const preferences = flow( [
 
 		return state;
 	},
-	pinnedPluginItems( state, action ) {
```
Comment (gziolo) [inline packages/edit-post/src/store/reducer.js]: This is going to remove the list of pinned/unpinned items stored in the browser for users.

File: packages/interface/src/store/defaults.js [hunk 2]
```diff
@@ -0,0 +1,14 @@
+export const DEFAULTS = {
+	areaControl: {
+		singleActiveAreas: {
```
Comment (gziolo) [inline packages/interface/src/store/defaults.js]: Can you explain what's the difference between `singleActiveAreas` and `multipleActiveAreas `? Can't it be merged into one object?

File: packages/edit-post/src/components/header/index.js [hunk 3]
```diff
@@ -94,7 +94,7 @@ function Header() {
 					aria-expanded={ isEditorSidebarOpened }
 					shortcut={ shortcut }
 				/>
-				<PinnedPlugins.Slot />
+				<PinnedItems.Slot scope="edit-post/complementary-area" />
```
Comment (gziolo) [inline packages/edit-post/src/components/header/index.js]: It should list all types of pinned items for a given page, it was never meant to work only with sidebar plugins:

```suggestion
				<PinnedItems.Slot scope="edit-post" />
```

File: packages/interface/package.json [hunk 4]
```diff
@@ -0,0 +1,37 @@
+{
+	"name": "@wordpress/interface",
+	"version": "0.0.1",
+	"private": true,
```
Comment (gziolo) [inline packages/interface/package.json]: As long as public packages depend on it, it can't be private.

File: packages/interface/package.json [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+	"module": "build-module/index.js",
+	"react-native": "src/index",
+	"dependencies": {
+		"@babel/runtime": "^7.8.3",
+		"@wordpress/components": "../components",
+		"@wordpress/data": "4.14.0",
+		"@wordpress/element": "../element",
+		"@wordpress/i18n": "../i18n",
+		"@wordpress/icons": "../icons",
+		"@wordpress/plugins": "../plugins",
```
Comment (gziolo) [inline packages/interface/package.json]: Dependencies should be declared as follows:

```suggestion
		"@wordpress/components": "file:../components",
		"@wordpress/data": "file:../data",
		"@wordpress/element": "file:../element",
		"@wordpress/i18n": "file:../i18n",
		"@wordpress/icons": "file:../icons",
		"@wordpress/plugins": "file:../plugins",
```

File: packages/edit-post/src/store/reducer.js (hunk 1 above)
Comment (jorgefilipecosta) [reply packages/edit-post/src/store/reducer.js]: Yes, the current preferences are removed with this change. I think it is not a big issue.

File: packages/interface/package.json [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+		"classnames": "^2.2.5",
+		"lodash": "4.17.15"
```
Comment (gziolo) [inline packages/interface/package.json]: All external dependencies should use ranges:

```suggestion
		"lodash": "^4.17.15"
```

File: packages/interface/src/store/defaults.js (hunk 2 above)
Comment (jorgefilipecosta) [reply packages/interface/src/store/defaults.js]: The logic for singleActiveAreas ensures that only one area is kept active, when we enable an area the other area becomes disable.
multipleActiveAreas allows multiple active areas in a given zone, when we enable an area the other areas that were active are kept active.

File: packages/interface/src/components/plugin-complementary-area-header/index.js [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+ * WordPress dependencies
+ */
+import { __ } from '@wordpress/i18n';
+import { Button } from '@wordpress/components';
+import { close } from '@wordpress/icons';
+
+const PluginComplementaryAreaHeader = ( {
+	smallScreenTitle,
+	toggleComplementaryAreaShortcut,
+	closeComplementaryArea,
```
Comment (gziolo) [inline packages/interface/src/components/plugin-complementary-area-header/index.js]: How about using shorter names for those props like or something similar:
```suggestion
	toggleShortcut,
	onClose,
```

The usage `<PluginComplementaryAreaHeader onClose={ onClose } />` should have less load this way.

File: packages/components/src/higher-order/with-focus-return/index.js [hunk 8]
```diff
@@ -79,7 +79,7 @@ function withFocusReturn( options ) {
 				// the default behavior otherwise occurring here. This allows
 				// for support that the `onFocusReturn` decides to allow the
 				// default behavior to occur under some conditions.
-				if ( onFocusReturn() === false ) {
+				if ( onFocusReturn( this.props.childProps ) === false ) {
```
Comment (gziolo) [inline packages/components/src/higher-order/with-focus-return/index.js]: I can't find any references to this new prop. Does it fix some bugs?

File: packages/edit-post/src/components/layout/index.js [hunk 9]
```diff
@@ -141,7 +141,7 @@ function Layout() {
 									</div>
 								) }
 								<SettingsSidebar />
-								<Sidebar.Slot />
+								<PluginComplementaryArea.Slot scope="edit-post/complementary-area" />
```
Comment (gziolo) [inline packages/edit-post/src/components/layout/index.js]: This slot is has the `complementary-area` part already included in the name if the component, does it need to be repeated in the scope or is it enough to leave the name of the page? Like:
```suggestion
								<PluginComplementaryArea.Slot scope="edit-post" />
```

**Review by gziolo — COMMENTED** [review]
I added a few more comments to check. This PR became huge :)

File: packages/interface/src/store/defaults.js (hunk 2 above)
Comment (gziolo) [reply packages/interface/src/store/defaults.js]: Is `multipleActiveAreas` used only for pinned items? If it's the case then it probably can be simplified as pinned items should mix shortcuts to all types of plugins for a given page type `edit-site` vs `edit-post` vs `edit-widgets`.

Comment (jorgefilipecosta) [reply packages/interface/src/store/defaults.js]: For now, multipleActiveAreas is only used for pinned items but my plan was to use it for other things, like store which panels are open.

File: packages/interface/package.json (hunk 4 above)
Comment (jorgefilipecosta) [reply packages/interface/package.json]: Plugin packages are bundling it so I guess it can be private no?

File: packages/components/src/higher-order/with-focus-return/index.js (hunk 8 above)
Comment (jorgefilipecosta) [reply packages/components/src/higher-order/with-focus-return/index.js]: I think this change was not needed anymore It was part of an experiment and I missed the removal.

File: packages/interface/package.json (hunk 4 above)
Comment (chrisvanpatten) [reply packages/interface/package.json]: Would this mean third parties couldn't use it? Really eager for this :)

Comment (jorgefilipecosta) [reply packages/interface/package.json]: Yes, for now, third parties would not be able to use this package. But it should end up being public when it is expanded and we are confident with its API :)

**Comment by jorgefilipecosta** [comment]
Hi @gziolo, thank you for the review the PR was updated and rebased.

File: packages/interface/package.json (hunk 4 above)
Comment (youknowriad) [reply packages/interface/package.json]: I don't see why this package should be private if it's bundled. third-parties should be able to consume the package from npm and bundle it similarily to Gutenberg.

Comment (youknowriad) [reply packages/interface/package.json]: And breaking changes in npm packages are allowed. They are forbidden for WordPress global APIs

Comment (gziolo) [reply packages/interface/package.json]: npm doesn't support bundling, it's only bundled from the webpack perspective or more precisely it isn't listed as an entry point. `@wordpress/edit-post` is a public package and `@wordpress/interface` is listed as a dependency so it needs to be exposed so people could install it.

File: packages/edit-post/src/components/sidebar/plugin-sidebar/index.js [hunk 10]
```diff
@@ -130,44 +67,26 @@ function PluginSidebar( props ) {
...
-	withPluginContext( ( context, ownProps ) => {
+export default function PluginSidebarEditPost( { className, ...props } ) {
+	const { postTitle, shortcut } = useSelect( ( select ) => {
 		return {
-			icon: ownProps.icon || context.icon,
-			sidebarName: `${ context.name }/${ ownProps.name }`,
+			postTitle: select( 'core/editor' ).getEditedPostAttribute(
+				'title'
+			),
+			shortcut: select(
```
Comment (gziolo) [inline packages/edit-post/src/components/sidebar/plugin-sidebar/index.js]: Should all sites share the same shortcut for toggling the sidebar/complementary area?

File: packages/edit-post/src/store/actions.js [hunk 11]
```diff
@@ -3,29 +3,34 @@
...
  *
- * @return {Object} Action object.
+ * @yield {Object} Action object.
  */
-export function closeGeneralSidebar() {
-	return {
-		type: 'CLOSE_GENERAL_SIDEBAR',
-	};
+export function* closeGeneralSidebar() {
+	yield* openGeneralSidebar();
```
Comment (gziolo) [inline packages/edit-post/src/store/actions.js]: Is it a bug, why would you call open to close?

File: packages/edit-site/src/components/sidebar/index.js [hunk 12]
```diff
@@ -1,20 +1,35 @@
...
-
 function Sidebar() {
 	return (
-		<div className="edit-site-sidebar">
-			<Panel header={ __( 'Inspector' ) }>
+		<>
+			<PluginComplementaryArea.Slot scope="edit-site" />
+			<PluginComplementaryArea
+				scope="edit-site"
+				complementaryAreaName="edit-site/block-inspector"
```
Comment (gziolo) [inline packages/edit-site/src/components/sidebar/index.js]: Would it work if we would simplify it a bit to:
```suggestion
				name="block-inspector"
```

`complementaryAreaName` could be combined as `scope` + `name`. What do you think?

File: packages/interface/package.json [hunk 13]
```diff
@@ -0,0 +1,37 @@
+{
+	"name": "@wordpress/interface",
+	"version": "0.0.1",
+	"private": true,
+	"description": "Interface module for WordPress.",
```
Comment (gziolo) [inline packages/interface/package.json]: Do you have some ideas on making this description more friendly for those using npm to find packages?

File: packages/interface/package.json [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+	"author": "The WordPress Contributors",
+	"license": "GPL-2.0-or-later",
+	"keywords": [
+		"wordpress"
```
Comment (gziolo) [inline packages/interface/package.json]: We can add more keywords, e.g.:
```suggestion
		"wordpress",
		"interface",
		"components"
```

File: packages/interface/src/components/index.js [hunk 15]
```diff
@@ -0,0 +1,3 @@
+export { default as PluginComplementaryArea } from './plugin-complementary-area';
+export { default as PluginComplementaryAreaHeader } from './plugin-complementary-area-header';
+export { default as PinnedItems } from './pinned-items';
```
Comment (gziolo) [inline packages/interface/src/components/index.js]: Now that is going to be exposed for public usage, should we prefix it with the plugin keyword? `PluginPinnedItems` or `PinnedPluginItems`?

**Review by gziolo — COMMENTED** [review]
Thanks for addressing the feedback shared. I took another pass, PR is huge :)

Do you plan to add unit tests for the store implementation?

File: packages/edit-post/src/components/layout/index.js [hunk 16]
```diff
@@ -145,7 +145,7 @@ function Layout() {
 									</div>
 								) }
 								<SettingsSidebar />
-								<Sidebar.Slot />
+								<PluginComplementaryArea.Slot scope="edit-post" />
```
Comment (youknowriad) [inline packages/edit-post/src/components/layout/index.js]: Can you explain the naming here? `PluginComplementaryArea`?

Comment (youknowriad) [reply packages/edit-post/src/components/layout/index.js]: Actually, what I'm missing in this PR is documentation, what are these components for? How to introduce plugins to a new screen? This would help personally understand the APIs and provide better feedback.

File: packages/edit-post/src/components/sidebar/plugin-sidebar/index.js (hunk 10 above)
Comment (jorgefilipecosta) [reply packages/edit-post/src/components/sidebar/plugin-sidebar/index.js]: I think sharing the same shortcut makes sense. I guess we can keep the shortcut prop but in the future add a default value. In our core use cases we would not pass the prop so we would make all the screen share the same shortcut.

File: packages/edit-post/src/store/actions.js (hunk 11 above)
Comment (jorgefilipecosta) [reply packages/edit-post/src/store/actions.js]: It is not a bug we are calling openGeneralSidebar without any sidebar name so the name is undefined and in that case, we will make sure no sidebar is available in that area.
But I will update the code, to not use openGeneralSidebar as it is confusing.

File: packages/edit-site/src/components/sidebar/index.js (hunk 12 above)
Comment (jorgefilipecosta) [reply packages/edit-site/src/components/sidebar/index.js]: In most cases, the name of the area is `${ context.name }/${ ownProps.name }`.
That means an area is named with the plugin id of the plugin that originated it and the name of the sidebar.
I'm not sure if we should include the scope as the name of the area, because in the store the area is already in a tree under the scope.

In the case for our own core sidebars, we are passing an explicit name I totally agree the name could be "block-inspector" instead of "edit-site/block-inspector", I used the later because in edit-post we were also using "edit-post/document" instead of just "document" and "edit-post/block" instead of "block". Renaming things in this PR will make it even bigger, as we would need back-compatibility for the previous names.

File: packages/interface/src/components/index.js (hunk 15 above)
Comment (jorgefilipecosta) [reply packages/interface/src/components/index.js]: Even for PluginComplementaryArea, and PluginComplementaryAreaHeader, I'm not sure we should include plugin prefix, we are using these components our selves in the core. Maybe we can just discard all the plugin prefixes?

Comment (gziolo) [reply packages/interface/src/components/index.js]: Right, it might work this way as well 👍

File: packages/interface/package.json (hunk 13 above)
Comment (jorgefilipecosta) [reply packages/interface/package.json]: I extended a little bit the description, but for now, this package does not offer much and will probably not be very useful to third-party users.
When the functionality of this package is expanded and we are confident with what the package offers I guess we can expand even more the description and make it more concrete.

File: packages/interface/src/components/index.js (hunk 15 above)
Comment (jorgefilipecosta) [reply packages/interface/src/components/index.js]: Names were updated.

File: packages/edit-post/src/components/layout/index.js (hunk 16 above)
Comment (jorgefilipecosta) [reply packages/edit-post/src/components/layout/index.js]: Hi @youknowriad, I added docs to the API we are exposing.
This PR is exposing two actions, two selectors, and two pairs of slot fill components.
Let me know If the documentation helped to understand how the mechanism we are adding work.

File: packages/interface/README.md [hunk 17]
```diff
@@ -0,0 +1,77 @@
...
+```
+
+_This package assumes that your code will run in an **ES2015+** environment. If you're using an environment that has limited or no support for ES2015+ such as lower versions of IE then using [core-js](https://github.com/zloirock/core-js) or [@babel/polyfill](https://babeljs.io/docs/en/next/babel-polyfill) will add support for these methods. Learn more about it in [Babel docs](https://babeljs.io/docs/en/next/caveats)._
+
+
+### API Usage
+
+
+
+#### Store
```
Comment (youknowriad) [inline packages/interface/README.md]: Reading the README it's unclear to me what this store is about at this point, I'm introduced to something called areas (multiple and single) but I don't know what these areas are at this point.

File: packages/interface/README.md [hunk 18]
```diff
@@ -0,0 +1,77 @@
...
+
+wp.data.select( 'core/interface' ).isMultipleActiveAreaActive( 'edit-post/complementary-area', 'edit-post/block-patterns-sidebar' ); -> true
+```
+
+
+### Components
+
+Four components are exposed. The components are based on slot & fill paradigm two components are "fills" and two components are "slots".
+
+`ComplementaryArea` and `ComplementaryArea.Slot` form a slot fill pair to render complementary areas. Multiple `ComplementaryArea` components representing different complementary areas may be rendered at the same time, but only one appears on the slot depending on which complementary area is enabled.
```
Comment (youknowriad) [inline packages/interface/README.md]: `ComplementaryArea` feels like a weird name to me. I don't have a better name though. Thoughts @aduth @mtias

File: packages/interface/README.md [hunk 19]
```diff
@@ -0,0 +1,77 @@
...
+_This package assumes that your code will run in an **ES2015+** environment. If you're using an environment that has limited or no support for ES2015+ such as lower versions of IE then using [core-js](https://github.com/zloirock/core-js) or [@babel/polyfill](https://babeljs.io/docs/en/next/babel-polyfill) will add support for these methods. Learn more about it in [Babel docs](https://babeljs.io/docs/en/next/caveats)._
+
+
+### API Usage
+
+
+
+#### Store
+
+#### Single Active Area Functionality
```
Comment (youknowriad) [inline packages/interface/README.md]: I'm not sure i understand why we need separate actions/selectors for multiple or single active areas. I feel this could be just a prop of the `ComplementaryArea.Slot` component.

Comment (youknowriad) [reply packages/interface/README.md]: Also, do we need both use-cases in the current screen?

**Review by youknowriad — COMMENTED** [review]
Thanks for the docs, it helps a lot.

File: packages/interface/README.md (hunk 17 above)
Comment (youknowriad) [reply packages/interface/README.md]: Technical remark: Adding a store to a package make that store a singleton, thus a bit less reusable. I'd have preferred if the API was only based on components but I'm not sure how possible is that.

File: packages/interface/README.md (hunk 18 above)
Comment (gziolo) [reply packages/interface/README.md]: `complementary` is one of the ARIA regions:
https://www.w3.org/TR/wai-aria-1.1/#region

> Authors SHOULD limit use of the region role to sections containing content with a purpose that is not accurately described by one of the other `landmark` roles, such as `main`, `complementary`, or `navigation`.

More details:

https://www.w3.org/TR/wai-aria-1.1/#complementary

https://www.w3.org/TR/wai-aria-practices/examples/landmarks/complementary.html

Comment (gziolo) [reply packages/interface/README.md]: For the full picture, it's based on previous discussions that using `Sidebar` as the component name is wrong here because it behaves differently on mobile web (modal like) and native mobile (bottom sheet). In addition, we should try to avoid naming high-level components after their placement on the screen but rather phrase them in a way that expresses their intent, some good examples that exist:
- `BlockControls`
- `InsepectorControls`
- `PinnedItems`

File: packages/edit-post/src/components/layout/index.js (hunk 16 above)
Comment (gziolo) [reply packages/edit-post/src/components/layout/index.js]: That's useful 👍

File: packages/interface/README.md (hunk 19 above)
Comment (jorgefilipecosta) [reply packages/interface/README.md]: Hi @youknowriad, the single reducer handles the case where only one or zero areas are active in a given scope, so enabling an area disables the previous active are. It is used to track which complementary area (sidebar) is active in a given screen. The multiple reducer handles cases, where multiple areas can be active in a given scope, enabling an area, does not impacts the other areas. It is used to track which pinned items are shown.

File: packages/interface/README.md (hunk 17 above)
Comment (jorgefilipecosta) [reply packages/interface/README.md]: > I'd have preferred if the API was only based on components but I'm not sure how possible is that.

We can have a provider that keeps state passes it by context and the ComplementaryArea and PinnedItems components access that state.
There are some challenges with that approach:
	- As it is not a store we would not automatically use the persistent mechanisms a store provides and we would would need to implement something new.
	- Currently to open/close sidebars third-party plugins can dispatch actions to edit-post (our code also uses that). We totally removed the reducers and store logic for edit-post sidebars but we are being back-compatible with all the edit-post store actions and selectors. If we used a components approach without a store being back-compatible would be more tricky.

Comment (gziolo) [reply packages/interface/README.md]: It'd be interesting to explore persistent contexts :)

Comment (jorgefilipecosta) [reply packages/interface/README.md]: > It'd be interesting to explore persistent contexts :)

That's is something we could do, the more tricky part would be the second point: being back-compatible with the current store calls.
The provider for this area would also need to be above the plugin content provider and we would need to provide an API probably context-based that plugins can use to open complementary areas in new screens (where they can not use the edit-post back-compatibility layer).

File: packages/interface/README.md (hunk 19 above)
Comment (youknowriad) [reply packages/interface/README.md]: Yes, I understand buy why we need separate actions, can we just use the same actions (like we do for selected blocks for instance and their actions). The "prop" for the slot could be responsible of adapting the behavior properly.

What I'm saying is that a consumer shouldn't have to care about whether a single or multiple areas can be enabled at the same time, this should be internal to the Slot. The consumer should just enable/disable areas.

File: packages/interface/README.md (hunk 17 above)
Comment (youknowriad) [reply packages/interface/README.md]: Seems like a bit too much to replicate, I wonder if at some point we could move away from Global stores and just have "local stores" (which is basically the same thing as a context)

File: packages/interface/README.md (hunk 19 above)
Comment (jorgefilipecosta) [reply packages/interface/README.md]: Hi @youknowriad, so essentially the idea is to remove logic from the store the store? The store would just contain active areas and would be UP for the store users to have the logic and know if multiple active areas can be enabled or not?


If that's the case there are some disadvantages we would for example not be able to open a sidebar by just dispatching an action. In order to open a sidebar, one would need first to check if there is an active sidebar if yes dispatch an action to disable that area and then dispatch an action to enable the new sidebar.

Comment (youknowriad) [reply packages/interface/README.md]: > so essentially the idea is to remove logic from the store the store? The store would just contain active areas and would be UP for the store users to have the logic and know if multiple active areas can be enabled or not?

Not really

> If that's the case there are some disadvantages we would for example not be able to open a sidebar by just dispatching an action

this should be a single action still but since the complementary slot has a config saying that it's a single active slot. It will automatically do its magic to disable the old one.

What I'm saying is that a consumer should do just `activateComplementaryArea( name )` (random naming here) regardless of whether the slot will display one or multiple at the same time. It's the responsibility of the implementation to know how it's configured, it shouldn't be the responsibility of the caller because otherwise for the same slot, some folks will say it's single while others will say it's multi

**Review by youknowriad — APPROVED** [review]
I like this direction. Thanks for the updates.

File: packages/edit-site/src/components/sidebar/index.js [hunk 20]
```diff
@@ -1,20 +1,35 @@
...
 import { __ } from '@wordpress/i18n';
+import { cog, pencil } from '@wordpress/icons';
 
 const { Slot: InspectorSlot, Fill: InspectorFill } = createSlotFill(
 	'EditSiteSidebarInspector'
 );
-
 function Sidebar() {
 	return (
-		<div className="edit-site-sidebar">
```
Comment (nosolosw) [inline packages/edit-site/src/components/sidebar/index.js]: @jorgefilipecosta want to check why we removed this, because I'm adding it back at https://github.com/WordPress/gutenberg/pull/20530/files?file-filters%5B%5D=.js#diff-3924a41ad88da62082971d8c0aa8b20fL14

The style.css is still targeting that class. I've only noticed after the changes made to the hooks (color, line-height), because some styles were leaking to the edit-site sidebar.

File: packages/edit-site/src/components/header/index.js [hunk 21]
```diff
@@ -62,7 +62,7 @@ export default function Header() {
 			</div>
 			<div className="edit-site-header__actions">
 				<SaveButton />
-				<MoreMenu />
```
Comment (vindl) [inline packages/edit-site/src/components/header/index.js]: Was this removal intentional or did it just end up there during the rebase?

Comment (youknowriad) [reply packages/edit-site/src/components/header/index.js]: Probably a rebase issue.

**Comment by youknowriad** [comment]
@vindl @jorgefilipecosta Should we move EditorSkeleton and Fullscreen to the new package?

**Comment by gziolo** [comment]
Just wanted to highlight my comments that weren't addressed:

- https://github.com/WordPress/gutenberg/pull/20698/files#r392245168 - new npm package can't be marked as private
- unit tests for old store logic were removed, there are no tests for the new store

I don't remember if I mentioned it but users will lost their saved preferences for pinned items because the migration path is missing. It was persisted in the local storage in the `core/edit-post` store so far.

File: packages/edit-site/src/components/header/index.js (hunk 21 above)
Comment (vindl) [reply packages/edit-site/src/components/header/index.js]: > Probably a rebase issue.

Ok, I'll spin up a quick PR to bring it back.

**Comment by vindl** [comment]
> @vindl @jorgefilipecosta Should we move EditorSkeleton and Fullscreen to the new package?

As for the Fullscreen - yes, I think so and I can look into this soon. As for the EditorSkeleton, I'd have to check its code a bit more before I can say for sute.

**Comment by vindl** [comment]
> @vindl @jorgefilipecosta Should we move EditorSkeleton and Fullscreen to the new package?

@youknowriad I started the PRs for moving those in: https://github.com/WordPress/gutenberg/pull/21334 and https://github.com/WordPress/gutenberg/pull/21335.

**Comment by aduth** [comment]
Note: I added the `interface` as an unchecked task as part of #18838, since it was not included with the introduction of the package.

**Comment by gziolo** [comment]
> * https://github.com/WordPress/gutenberg/pull/20698/files#r392245168 - new npm package can't be marked as private

Addressed in https://github.com/WordPress/gutenberg/pull/21417.

**Comment by jorgefilipecosta** [comment]
> unit tests for old store logic were removed, there are no tests for the new store

Hi @gziolo, I will soon submit a set of unit tests the functionally of this package. I just want to understand the shape the package will follow as things can still change so we don't waste effort on tests that may need updates.

> I don't remember if I mentioned it but users will lost their saved preferences for pinned items because the migration path is missing. It was persisted in the local storage in the core/edit-post store so far.

I think it was discussed before, personally I think losing the pinned items is not a big issue as users can easily update them, add the migration path may make our code more complex. But if you all think we should have some migration I can propose something. Any thoughts on this @gziolo, @youknowriad?

File: packages/interface/src/components/complementary-area/index.js [hunk 22]
```diff
@@ -0,0 +1,147 @@
...
+		},
+		[ complementaryAreaIdentifier, scope ]
+	);
+	const { enableComplementaryArea, disableComplementaryArea } = useDispatch(
+		'core/interface'
+	);
+	const { pinItem, unpinItem } = useDispatch( 'core/interface' );
+	return (
+		<>
+			{ isPinned && (
```
Comment (gziolo) [inline packages/interface/src/components/complementary-area/index.js:70]: The fact that `isPinned` is now outside of the `<PinnedItems />` fill created a regression as described in #28546 =.
